src/entry_point.rs

```diff
@@ -61,29 +85,83 @@ pub unsafe extern "C" fn _start(
         b .Lyield_loop
 
         .Lstack_init:
-        // Initialize the stack pointer. The stack pointer is computed as
+        // Compute the stacktop (stack_start). The stacktop is computed as
         // stack_size + mem_start plus padding to align the stack to a multiple
         // of 8 bytes. The 8 byte alignment is to follow ARM AAPCS:
         // http://infocenter.arm.com/help/index.jsp?topic=/com.arm.doc.faqs/ka4127.html
-        ldr ip, [r0, #36]  // ip = text_start->stack_size
-        add ip, ip, r1     // ip = text_start->stack_size + mem_start
-        add ip, #7         // ip = text_start->stack_size + mem_start + 7
-        bic r1, ip, #7     // r1 = (text_start->stack_size + mem_start + 7) & ~0x7
-        mov sp, r1         // sp = r1
-
-        // Call rust_start. text_start, stack_top, and app_heap_break are
-        // already in the correct registers.
+        ldr r4, [r0, #36]  // r4 = app_start->stack_size
+        add r4, r4, r1     // r4 = app_start->stack_size + mem_start
+        add r4, #7         // r4 = app_start->stack_size + mem_start + 7
+        bic r4, r4, #7     // r4 = (app_start->stack_size + mem_start + 7) & ~0x7
+        mov sp, r4         // sp = r4
+
+        // We need to pass app_start, stacktop and app_heap_break to rust_start.
+        // Temporarily store them in r6, r7 and r8
+        mov r6, r0
+        mov r7, sp
+
+        // Debug support, tell the kernel the stack location
+        //
+        // memop(10, stacktop)
+        // r7 contains stacktop
+        mov r0, #10
+        mov r1, r7
+        svc 4
+
+        // Debug support, tell the kernel the heap_start location
+        mov r0, r6
+        ldr r4, [r0, #24] // r4 = app_start->bss_start
+        ldr r5, [r0, #28] // r5 = app_start->bss_size
+        add r4, r4, r5    // r4 = bss_start + bss_size
+        //
+        // memop(11, r4)
+        mov r0, #11
+        mov r1, r4
+        svc 4
+
+        // Store heap_start (and soon to be app_heap_break) in r8
+        mov r8, r4
+
+        // There is a possibility that stack + .data + .bss is greater than
+        // 3072. Therefore setup the initial app_heap_break to heap_start (that
+        // is zero initial heap) and let rust_start determine where the actual
+        // app_heap_break should go.
+        //
+        // Also, because app_heap_break is where the unprivileged MPU region
+        // ends, in case mem_start + stack + .data + .bss is greater than
+        // initial app_heap_break (mem_start + 3072), we will get a memory fault
+        // in rust_start when initializing .data and .bss. Setting
+        // app_heap_break to heap_start avoids that.
+
+        // memop(0, r8)
+        mov r0, #0
+        mov r1, r8
+        svc 4
```

Comment:
Why did you move the memop() calls into assembly?

Reply:
The main motivation for moving memop calls into assembly was to make sure that by the time we got to `rust_start` we are in a state where we can get correct PC in case of a fault. This turned out to be important when debugging and trying to understand MPU code.

Reply:
I think the kernel panic-ing before this is a kernel bug and shouldn't be worked around in userspace. IIRC Tock is supposed to prevent apps from DoS-ing each other completely, and a kernel panic is a DoS vulnerability.

That said, I suppose I'm okay with this as long as the plan is to correct it once the kernel bug is fixed.

Reply:
Would you mind linking the Tock kernel bug here for future reference? If not, would you mind creating one w/ a reproduction case?

Reply:
Is it really a kernel panic? Or the process faulting (which by default results in a kernel panic rather than restarting or killing the process for development)? If it's the former, yes, please point to the bug (we should fix it!)